File: packages/py-build/scripts/publish.mjs
#!/usr/bin/env node
// Publish the staged py artifacts to R2 under their build-lock hash prefix.
//
//   node scripts/publish.mjs [--local|--remote] [--dry-run]   (--local default)
//
// Local seeds the dev miniflare R2 simulator (pnpm py:seed); remote publishes
// to the real avlo-py bucket. Invariant enforced by the preflight: RESTAGE ⇒
// RESEED — every uploaded byte is re-hashed against the committed build-lock,
// so a stale dist/ or a stale lock refuses loudly instead of serving a mix.
//
// Upload plan: every lock artifact + bundle tar, each with its .br sibling,
// then manifest.json strictly LAST — an aborted upload leaves an incomplete
// prefix WITHOUT its completion marker, which the probe treats as unpublished.
import { execFileSync } from 'node:child_process';
import { createHash } from 'node:crypto';
import { existsSync, readFileSync, statSync } from 'node:fs';
import { dirname, join, resolve } from 'node:path';
import { fileURLToPath } from 'node:url';

const pkgRoot = resolve(dirname(fileURLToPath(import.meta.url)), '..');
const repoRoot = resolve(pkgRoot, '../..');
const args = process.argv.slice(2);
const remote = args.includes('--remote');
const dryRun = args.includes('--dry-run');
const BUCKET = 'avlo-py';

const lock = JSON.parse(readFileSync(join(repoRoot, 'packages/py-loader/build-lock.json'), 'utf8'));
const sha256 = (buf) => createHash('sha256').update(buf).digest('hex');

// Raw source locations (the .br siblings live beside them — they never reach fork/).
const SOURCES = {
  'pyodide.asm.js': 'dist/raw/pyodide.asm.js',
  'pyodide.asm.wasm': 'dist/raw/pyodide.asm.wasm',
  'pyodide.mjs': 'dist/raw/pyodide.mjs',
  'python_stdlib.zip': 'dist/stage/python_stdlib.zip',
  'baseline.snap': 'dist/baseline.snap',
};
// NOTE: contentType has no fallback — an unmapped extension would pass the
// literal string "undefined" to `wrangler --content-type`.
const MIME = {
  '.mjs': 'text/javascript',
  '.js': 'text/javascript',
  '.wasm': 'application/wasm',
  '.zip': 'application/zip',
  '.json': 'application/json',
  '.tar': 'application/x-tar',
  '.snap': 'application/octet-stream',
};
const contentType = (file) => MIME[file.slice(file.lastIndexOf('.'))];

// ---- preflight: restage ⇒ reseed ---------------------------------------------
const uploads = []; // { key, path, contentType, encoding? }
const preflight = (name, rel, want) => {
  const p = join(pkgRoot, rel);
  if (!existsSync(p)) {
    console.error(`missing ${rel} — build/stage first`);
    process.exit(1);
  }
  const buf = readFileSync(p);
  if (buf.length !== want.size || sha256(buf) !== want.sha256) {
    console.error(`${rel} drifted from build-lock (restage ⇒ reseed: rerun stage.mjs, commit the lock, reseed)`);
    process.exit(1);
  }
  const brPath = `${p}.br`;
  if (!existsSync(brPath)) {
    console.error(`missing ${rel}.br — run compress.mjs`);
    process.exit(1);
  }
  if (statSync(brPath).mtimeMs < statSync(p).mtimeMs) {
    console.error(`${rel}.br is OLDER than its source — run compress.mjs`);
    process.exit(1);
  }
  const keyBase = name.endsWith('.tar') ? `${lock.buildHash}/bundles/${name}` : `${lock.buildHash}/${name}`;
  uploads.push({ key: keyBase, path: p, contentType: contentType(name) });
  uploads.push({ key: `${keyBase}.br`, path: brPath, contentType: contentType(name), encoding: 'br' });
};
for (const [name, rel] of Object.entries(SOURCES)) preflight(name, rel, lock.artifacts[name]);
for (const [b, want] of Object.entries(lock.bundles)) preflight(`${b}.tar`, `dist/stage/bundles/${b}.tar`, want);

const manifestPath = join(repoRoot, 'web/public/py-dev/fork/manifest.json');
if (!existsSync(manifestPath)) {
  console.error('staged manifest.json missing — run stage.mjs');
  process.exit(1);
}
const manifestBytes = readFileSync(manifestPath);
if (JSON.parse(manifestBytes.toString('utf8')).buildHash !== lock.buildHash) {
  console.error('staged manifest.json buildHash != build-lock — rerun stage.mjs');
  process.exit(1);
}
uploads.push({ key: `${lock.buildHash}/manifest.json`, path: manifestPath, contentType: 'application/json' });

// ---- wrangler invocation -------------------------------------------------------
// --local: --persist-to must be ABSOLUTE and must NOT include the trailing /v3 —
// wrangler appends it, landing in the exact tree dev-miniflare reads
// (.wrangler/state/v3). --remote: config supplies the account context.
const modeArgs = remote
  ? ['--remote', '--config', join(repoRoot, 'workers/py/wrangler.jsonc')]
  : ['--local', '--persist-to', join(repoRoot, '.wrangler/state')];
const wr = (cmd, opts = {}) => execFileSync('npx', ['wrangler', 'r2', 'object', ...cmd, ...modeArgs], { cwd: repoRoot, ...opts });

if (remote && !dryRun) {
  // Divergence refusal: one probe of the completion marker.
  let existing = null;
  try {
    existing = wr(['get', `${BUCKET}/${lock.buildHash}/manifest.json`, '--pipe'], { stdio: ['ignore', 'pipe', 'pipe'] });
  } catch {
    /* absent → publish */
  }
  if (existing !== null) {
    if (Buffer.compare(existing, manifestBytes) === 0) {
      console.log(`already published: ${lock.buildHash} (manifest byte-identical)`);
      process.exit(0);
    }
    console.error(`HASH BUG: ${lock.buildHash}/manifest.json exists remotely with DIFFERENT bytes — same hash must mean same artifacts`);
    process.exit(1);
  }
}

for (const u of uploads) {
  const flags = ['--file', u.path, '--content-type', u.contentType, ...(u.encoding ? ['--content-encoding', u.encoding] : [])];
  console.log(`${dryRun ? '[dry-run] ' : ''}put ${BUCKET}/${u.key}${u.encoding ? ' (br)' : ''}`);
  if (!dryRun) wr(['put', `${BUCKET}/${u.key}`, ...flags], { stdio: ['ignore', 'ignore', 'inherit'] });
}
console.log(`${dryRun ? '[dry-run] ' : ''}published ${uploads.length} keys under ${lock.buildHash} (${remote ? 'remote' : 'local'})`);

File: packages/py-build/src/avlo_py_build/publish.py
"""Publish the staged py artifacts to R2 under their build-lock hash prefix.

  avlo-build publish [--local|--remote] [--dry-run]   (--local default)

Local seeds the dev miniflare R2 simulator (pnpm py:seed); remote publishes
to the real avlo-py bucket. Invariant enforced by the preflight: RESTAGE ⇒
RESEED — every uploaded byte is re-hashed against the committed build-lock,
so a stale dist/ or a stale lock refuses loudly instead of serving a mix.

Upload plan: every lock artifact + bundle tar, each with its .br sibling,
then manifest.json strictly LAST — an aborted upload leaves an incomplete
prefix WITHOUT its completion marker, which the probe treats as unpublished.

Transport is still `wrangler r2 object put` (no checksum flag — the
binding-put upgrade with server-side sha256 verify is researched in NOTES
Open items and lands with the publish/serving phase, likely as boto3 to the
S3 endpoint per the replatform plan §2.8).
"""

import hashlib
import json
import subprocess
import sys

from .paths import BUILD_LOCK, FORK_PUBLIC, PKG_ROOT, REPO_ROOT

BUCKET = "avlo-py"

# Raw source locations (the .br siblings live beside them — they never reach fork/).
SOURCES = {
    "pyodide.asm.mjs": "dist/raw/pyodide.asm.mjs",
    "pyodide.asm.wasm": "dist/raw/pyodide.asm.wasm",
    "pyodide.mjs": "dist/raw/pyodide.mjs",
    "python_stdlib.zip": "dist/stage/python_stdlib.zip",
}
# NOTE: no fallback — an unmapped extension must fail loudly, not upload a
# bogus content type.
MIME = {
    ".mjs": "text/javascript",
    ".wasm": "application/wasm",
    ".zip": "application/zip",
    ".json": "application/json",
    ".tar": "application/x-tar",
}


def _content_type(file: str) -> str:
    return MIME[file[file.rfind(".") :]]


def run(args) -> int:
    remote = args.remote
    dry_run = args.dry_run
    lock = json.loads(BUILD_LOCK.read_text())

    # ---- preflight: restage ⇒ reseed -----------------------------------------
    uploads: list[dict] = []  # { key, path, contentType, encoding? }

    def preflight(name: str, rel: str, want: dict) -> None:
        p = PKG_ROOT / rel
        if not p.exists():
            sys.exit(f"missing {rel} — build/stage first")
        buf = p.read_bytes()
        if len(buf) != want["size"] or hashlib.sha256(buf).hexdigest() != want["sha256"]:
            sys.exit(f"{rel} drifted from build-lock (restage ⇒ reseed: rerun avlo-build stage, commit the lock, reseed)")
        br_path = p.with_name(p.name + ".br")
        if not br_path.exists():
            sys.exit(f"missing {rel}.br — run avlo-build compress")
        if br_path.stat().st_mtime < p.stat().st_mtime:
            sys.exit(f"{rel}.br is OLDER than its source — run avlo-build compress")
        key_base = f"{lock['buildHash']}/bundles/{name}" if name.endswith(".tar") else f"{lock['buildHash']}/{name}"
        uploads.append({"key": key_base, "path": p, "contentType": _content_type(name)})
        uploads.append({"key": f"{key_base}.br", "path": br_path, "contentType": _content_type(name), "encoding": "br"})

    for name, rel in SOURCES.items():
        preflight(name, rel, lock["artifacts"][name])
    for b, want in lock["bundles"].items():
        preflight(f"{b}.tar", f"dist/stage/bundles/{b}.tar", want)

    manifest_path = FORK_PUBLIC / "manifest.json"
    if not manifest_path.exists():
        sys.exit("staged manifest.json missing — run avlo-build stage")
    manifest_bytes = manifest_path.read_bytes()
    if json.loads(manifest_bytes)["buildHash"] != lock["buildHash"]:
        sys.exit("staged manifest.json buildHash != build-lock — rerun avlo-build stage")
    uploads.append({"key": f"{lock['buildHash']}/manifest.json", "path": manifest_path, "contentType": "application/json"})

    # ---- wrangler invocation ----------------------------------------------------
    # --local: --persist-to must be ABSOLUTE and must NOT include the trailing
    # /v3 — wrangler appends it, landing in the exact tree dev-miniflare reads
    # (.wrangler/state/v3). --remote: config supplies the account context.
    mode_args = (
        ["--remote", "--config", str(REPO_ROOT / "workers/py/wrangler.jsonc")]
        if remote
        else ["--local", "--persist-to", str(REPO_ROOT / ".wrangler/state")]
    )

    def wr(cmd: list[str], **kwargs) -> subprocess.CompletedProcess:
        return subprocess.run(["npx", "wrangler", "r2", "object", *cmd, *mode_args], cwd=REPO_ROOT, **kwargs)

    if remote and not dry_run:
        # Divergence refusal: one probe of the completion marker.
        probe = wr(
            ["get", f"{BUCKET}/{lock['buildHash']}/manifest.json", "--pipe"],
            stdin=subprocess.DEVNULL,
            capture_output=True,
        )
        if probe.returncode == 0:
            if probe.stdout == manifest_bytes:
                print(f"already published: {lock['buildHash']} (manifest byte-identical)")
                return 0
            sys.exit(
                f"HASH BUG: {lock['buildHash']}/manifest.json exists remotely with DIFFERENT bytes — "
                "same hash must mean same artifacts"
            )

    for u in uploads:
        flags = ["--file", str(u["path"]), "--content-type", u["contentType"]]
        if u.get("encoding"):
            flags += ["--content-encoding", u["encoding"]]
        print(f"{'[dry-run] ' if dry_run else ''}put {BUCKET}/{u['key']}{' (br)' if u.get('encoding') else ''}")
        if not dry_run:
            r = wr(["put", f"{BUCKET}/{u['key']}", *flags], stdin=subprocess.DEVNULL, stdout=subprocess.DEVNULL)
            if r.returncode != 0:
                sys.exit(f"wrangler put failed for {u['key']}")
    print(f"{'[dry-run] ' if dry_run else ''}published {len(uploads)} keys under {lock['buildHash']} ({'remote' if remote else 'local'})")
    return 0
